File: src/main/java/com/codeperfection/shipit/service/shipping/placer/KnapsackPlacer.java
package com.codeperfection.shipit.service.shipping.placer;

import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;

@Component
public class KnapsackPlacer {

    public ItemsStorage place(Item[] items, int capacity) {
        final int[][] maxTotalPrices = calculateMaxTotalPrices(items, capacity);
        int maxTotalPrice = maxTotalPrices[items.length][capacity];
        int currentCapacity = capacity;
        List<Item> optimalPlacementItems = new ArrayList<>();

        for (int itemsSize = items.length; itemsSize > 0 && maxTotalPrice > 0; itemsSize--) {
            // If the max total price came from maxTotalPrices[itemsSize - 1][currentCapacity], then the last item
            // was not included to maximize the price. Otherwise the last item was included and we should add it to
            // the final result.
            if (maxTotalPrice != maxTotalPrices[itemsSize - 1][currentCapacity]) {
                final var lastItem = items[itemsSize - 1];
                optimalPlacementItems.add(lastItem);
                maxTotalPrice -= lastItem.getPrice();
                currentCapacity -= lastItem.getVolume();
            }
        }
        return new Knapsack(capacity, maxTotalPrices[items.length][capacity], optimalPlacementItems);
    }

    private int[][] calculateMaxTotalPrices(Item[] items, int capacity) {
        final int[][] maxTotalPrices = new int[items.length + 1][capacity + 1];
        for (int currentCapacity = 0; currentCapacity <= capacity; currentCapacity++) {
            maxTotalPrices[0][currentCapacity] = 0;
        }

        for (int itemsSize = 1; itemsSize <= items.length; itemsSize++) {
            for (int currentCapacity = 0; currentCapacity <= capacity; currentCapacity++) {
                final int lastItemIndex = itemsSize - 1;
                final var lastItem = items[lastItemIndex];
                final int withoutLastItem = maxTotalPrices[lastItemIndex][currentCapacity];
                if (lastItem.getVolume() > currentCapacity) {
                    maxTotalPrices[itemsSize][currentCapacity] = withoutLastItem;
                } else {
                    final int capacityWithoutLastItem = currentCapacity - lastItem.getVolume();
                    final int withLastItem = lastItem.getPrice() +
                            maxTotalPrices[lastItemIndex][capacityWithoutLastItem];
                    maxTotalPrices[itemsSize][currentCapacity] = Math.max(withoutLastItem, withLastItem);
                }
            }
        }
        return maxTotalPrices;
    }
}

File: src/main/kotlin/com/codeperfection/shipit/service/shipping/placer/KnapsackPlacer.kt
package com.codeperfection.shipit.service.shipping.placer

import org.springframework.stereotype.Component
import java.util.*
import kotlin.math.max

data class Knapsack(val capacity: Int, val totalPrice: Int, val items: List<Item>) {

    data class Item(val parentId: UUID, val volume: Int, val price: Int)

    fun isEmpty(): Boolean = items.isEmpty()
}

@Component
class KnapsackPlacer {

    fun place(items: Array<Knapsack.Item>, capacity: Int): Knapsack {
        val maxTotalPrices = calculateMaxTotalPrices(items, capacity)
        return backtrackItemsUsedInKnapsack(items, capacity, maxTotalPrices)
    }

    /**
     * Calculate combination of items, which doesn't exceed capacity and have maximum possible price sum.
     */
    private fun calculateMaxTotalPrices(items: Array<Knapsack.Item>, capacity: Int): Array<IntArray> {
        val maxTotalPrices = Array(items.size + 1) { IntArray(capacity + 1) { 0 } }

        for (size in 1..items.size) {
            for (currentCapacity in 0..capacity) {
                maxTotalPrices[size][currentCapacity] = calculateMaxTotalPrice(
                    items = items,
                    size = size,
                    capacity = currentCapacity,
                    previousMaxTotalPrices = maxTotalPrices
                )
            }
        }
        return maxTotalPrices
    }

    private fun calculateMaxTotalPrice(
        items: Array<Knapsack.Item>,
        size: Int,
        capacity: Int,
        previousMaxTotalPrices: Array<IntArray>
    ): Int {
        val lastItem = items[size - 1]

        val withoutLastItem = previousMaxTotalPrices[size - 1][capacity]
        if (lastItem.volume > capacity) {
            // last item cannot be placed as it will exceed overall capacity
            return withoutLastItem
        }

        val capacityWithoutLastItem = capacity - lastItem.volume
        val withLastItem = lastItem.price + previousMaxTotalPrices[size - 1][capacityWithoutLastItem]
        return max(withoutLastItem, withLastItem)
    }

    private fun backtrackItemsUsedInKnapsack(
        items: Array<Knapsack.Item>,
        capacity: Int,
        maxTotalPrices: Array<IntArray>
    ): Knapsack {
        /**
         * After calculating max total possible prices for each combination of items and capacity
         * now we need to go backwards in order to retrieve actual items which make the calculated overall price.
         * In each step we need to consider item to be added to the knapsack or ignored,
         * based on whether the item was used in calculation of the goal max total price.
         * If item was involved in the max total price calculation
         * - we store item in the final item list
         * - adjust goal price and capacity to not consider attributes of the item
         * - move to next item
         */
        var goalPriceToFill = maxTotalPrices[items.size][capacity]
        var goalCapacityToFill = capacity

        val resultItems = mutableListOf<Knapsack.Item>()

        for (itemsSize in items.size downTo 1) {
            // Stop if we have collected all items summing up to the goal
            if (goalPriceToFill == 0) {
                break
            }

            /**
             * If the goal price came from maxTotalPrices[itemsSize - 1][ goalCapacityToFill ],
             * then the last item was not included to maximize the price.
             */
            if (goalPriceToFill == maxTotalPrices[itemsSize - 1][goalCapacityToFill]) {
                continue
            }

            val lastItem = items[itemsSize - 1]
            resultItems.add(lastItem)
            goalPriceToFill -= lastItem.price
            goalCapacityToFill -= lastItem.volume
        }

        return Knapsack(capacity, maxTotalPrices[items.size][capacity], resultItems)
    }
}
